logstash/class-logger.php

```diff
@@ -0,0 +1,426 @@
+			return; // Failed validation.
+
+		} elseif ( empty( $data['message'] ) || ! is_string( $data['message'] ) ) {
+			trigger_error( 'Missing required `message` in call to ' . esc_html( __METHOD__ ) . '(). Must be a string.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( strlen( $data['message'] ) > static::MAX_ENTRY_MESSAGE_SIZE ) {
+			trigger_error( 'Invalid `message` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_MESSAGE_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_id'] ) && ! is_int( $data['user_id'] ) || $data['user_id'] < 0 ) {
+			trigger_error( 'Invalid `user_id` in call to ' . esc_html( __METHOD__ ) . '(). Must be an integer >= 0.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_ua'] ) && ! is_string( $data['user_ua'] ) ) {
+			trigger_error( 'Invalid `user_ua` in call to ' . esc_html( __METHOD__ ) . '(). Must be a string.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['user_ua'] ) && strlen( $data['user_ua'] ) > static::MAX_ENTRY_USER_UA_SIZE ) {
+			trigger_error( 'Invalid `user_ua` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_USER_UA_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['extra'] ) && ! is_array( $data['extra'] ) && ! is_object( $data['extra'] ) && ! is_scalar( $data['extra'] ) ) {
+			trigger_error( 'Invalid `extra` in call to ' . esc_html( __METHOD__ ) . '(). Must be an ojbect, array, or scalar value.', E_USER_WARNING );
+			return; // Failed validation.
+
+		} elseif ( isset( $data['extra'] ) && strlen( wp_json_encode( $data['extra'] ) ) > static::MAX_ENTRY_EXTRA_SIZE ) {
+			trigger_error( 'Invalid `extra` in call to ' . esc_html( __METHOD__ ) . '(). Must be ' . esc_html( static::MAX_ENTRY_EXTRA_SIZE ) . ' bytes or less.', E_USER_WARNING );
+			return; // Failed validation.
+		}
+
+		// Adjust `feature` and JSON-encode pretty-print data.
+		$data['feature']        = 'a8c_vip_' . $data['feature'];
+
+		// Adjust `extra` and JSON-encode data that gets sent in the API request.
+		$data['extra'] = isset( $data['extra'] ) ? wp_json_encode( $data['extra'], JSON_PRETTY_PRINT ) : '';
+
+		// Adds a new log entry.
+		static::$entries[] = $data;
+
+		// Sends data to logstash on shutdown.
+		if ( ! has_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] ) ) {
+			add_action( 'shutdown', [ static::class, 'process_entries_on_shutdown' ] );
+		}
+	}
+
+	/**
+	 * Sends data to logstash via REST API on shutdown, viewable in Kibana.
+	 *
+	 * @since 2020-01-10
+	 */
+	public static function process_entries_on_shutdown() : void {
+		if ( static::$processed_entries ) {
+			return; // Already done.
+		}
+		static::$processed_entries = true;
```

Comment:
To ensure we don't slow down page load by the time it takes for the API requests, we could send the response data to the client with `fastcgi_finish_request()` before delivering the logs.
```suggestion
		static::$processed_entries = true;

		if ( function_exists( 'fastcgi_finish_request' ) ) {
			fastcgi_finish_request();
		}
```

This will probably need to set a high number as a priority for the `shutdown` hook to ensure this code is executed at last. Or we could try utilizing the native `register_shutdown_function()` instead of the WP hook.

Reply:
Good idea! Changed in 9d673368bd6ca699ed23e3ee230f9e77730e60df